src/engine/renderer/tr_shader.cpp

```diff
@@ -5992,209 +5990,188 @@ A second parameter will cause it to print in sorted order
+	stageTypeStrings[stageType_t::ST_COLLAPSE_reflection_CB] = "REFLECTION_CB";
+	stageTypeStrings[stageType_t::ST_ATTENUATIONMAP_XY] = "ATTENUATIONMAP_XY";
+	stageTypeStrings[stageType_t::ST_ATTENUATIONMAP_Z] = "ATTENUATIONMAP_XZ";
+
+	std::string stageTypeHeader = "stageType";
+	int maxStageTypeStringLen = stageTypeHeader.length();
+
+	for ( stageTypeStringMap::iterator iter = stageTypeStrings.begin();
+		iter !=stageTypeStrings.end(); iter++ )
+	{
+		maxStageTypeStringLen = std::max( maxStageTypeStringLen, (int)iter->second.length() );
+	}
+
+	std::string interactLightHeader = "interactLight";
+	int maxInteractLightStringLen = interactLightHeader.length();
+
+	std::string separator = " ";
+
+	std::string header = Str::Format(
+		"%-*s" "%s%-*s" "%s%-*s" "%s%-*s" "%s%s",
+		maxShaderTypeStringLen, shaderTypeHeader,
+		separator, maxShaderSortStringLen, shaderSortHeader,
+		separator, maxInteractLightStringLen, interactLightHeader,
+		separator, maxStageTypeStringLen, stageTypeHeader,
+		separator, "stageNumber:shaderName" );
+
+	std::string lineSeparator( header.length(), '-' );
+
+	Log::CommandInteractionMessage( lineSeparator );
+	Log::CommandInteractionMessage( header );
+	Log::CommandInteractionMessage( lineSeparator );
+
+	int stageCount = 0;
+	int highestStage = 0;
+
+	for ( int i = 0; i < tr.numShaders; i++ )
+	{
+		shader_t *shader = ri.Cmd_Argc() > 2 ? tr.sortedShaders[ i ] : tr.shaders[ i ];
 
 		if ( s && !Com_Filter( s, shader->name, false ) )
 		{
 			continue;
 		}
 
-		std::string str;
+		std::string foundShaderTypeString = "";
 
-		switch ( shader->type )
+		for ( shaderTypeStringMap::iterator iter = shaderTypeStrings.begin();
+			iter !=shaderTypeStrings.end(); iter++ )
 		{
-			case shaderType_t::SHADER_2D:
-				str += "2D   ";
-				break;
-
-			case shaderType_t::SHADER_3D_DYNAMIC:
-				str += "3D_D ";
+			if ( shader->type == iter->first )
+			{
+				foundShaderTypeString = iter->second;
 				break;
+			}
+		}
 
-			case shaderType_t::SHADER_3D_STATIC:
-				str += "3D_S ";
-				break;
+		std::string foundShaderSortString = "";
 
-			case shaderType_t::SHADER_LIGHT:
-				str += "ATTN ";
+		for ( shaderSortStringMap::iterator iter = shaderSortStrings.begin();
+			iter !=shaderSortStrings.end(); iter++ )
+		{
+			if ( shader->sort == Util::ordinal( iter->first ) )
+			{
+				foundShaderSortString = iter->second;
 				break;
+			}
 		}
 
-		// there may be multiple kind of collapse stage in one shader,
-		// display first one since there is only one column to tell about it
-		bool foundCollapseType = false;
-		for ( int i = 0; i < shader->numStages; i++ )
+		std::string foundInteractLightString = shader->interactLight ? "INTERACTLIGHT" : "";
+
+		std::string shaderNameString = shader->name;
+		shaderNameString += shader->defaultShader ? " (DEFAULTED)" : "";
+
+		for ( int j = 0; j < shader->numStages; j++, stageCount++ )
 		{
-			if ( shader->stages[ i ]->active )
+			highestStage = std::max( j, highestStage );
```

Comment:
Stage count stuff would be better updated outside this loop, based on `shader->numStages` rather than `j`.

Reply:
Done.